### PR TITLE
Client tests

### PR DESCRIPTION
**Pull request by Peterburnett**

### CODE REVIEW

File: tests/client_test.php [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+use enrol_oneroster\local\command;
+use enrol_oneroster\client_helper;
+use enrol_oneroster\local\endpoint;
+require_once(__DIR__ . '/../csv_data_helper.php');
+
+/**
+ * One Roster tests for the client_helper class.
+ *
+ * @package    enrol_oneroster
+ * @copyright  Andrew Nicols <andrew@nicols.co.uk>
```
Comment (Peterburnett) [inline tests/client_test.php:37]: copyright

File: tests/client_test.php [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+            $academicSessions,
+        );
+
+        $csvclient->synchronise();
+
+        $course = $DB->get_records('course');
+        $user = $DB->get_records('user');
+        $enrol = $DB->get_records('enrol');
+
+        $this->assertCount(3, $course);
```
Comment (Peterburnett) [inline tests/client_test.php:79]: For all of these, I want to assert the shape alongside the count. This ensures all the fields are aligned with the expected data, and not filled with junk or nulls

File: tests/client_test.php [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        $this->assertCount(2, $user);
+        $this->assertCount(8, $enrol);
+
+    }
+}
+
+
```
Comment (Peterburnett) [inline tests/client_test.php:86]: Its probably worth having tests around a bunch of your helper methods for data loading etc as well, ensuring that the files can be read correctly and error in cases where they should

Comment (Peterburnett) [reply tests/client_test.php:86]: and especially the validation functions. Probably use a DataProvider for these

https://github.com/catalyst/moodle-tool_mfa/blob/MOODLE_400_STABLE/tests/manager_test.php#L170

File: tests/fixtures/oneroster_client.php [hunk 4]
```diff
@@ -37,4 +38,5 @@
  */
 abstract class oneroster_client implements client_interface, rostering_client_interface {
     use root_oneroster_client;
+    use csv_oneroster_client;
```
Comment (Peterburnett) [inline tests/fixtures/oneroster_client.php:41]: Maybe this should be its own fixture so there aren't side effects on the Oauth classes tests

File: tests/processcsv_filecheck_test.php [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+/**
+ * One Roster tests for the client_helper class.
+ *
+ * @package    enrol_oneroster
+ * @copyright  Andrew Nicols <andrew@nicols.co.uk>
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ *
+ * @covers  enrol_oneroster\OneRosterHelper
+ */
+class processcsv_test extends \advanced_testcase {
```
Comment (Peterburnett) [inline tests/processcsv_filecheck_test.php:34]: CI checks if the test class name is aligned with the filename. Check this with coding style docs

File: tests/processcsv_filecheck_test.php [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+namespace enrol_oneroster\tests;
+use enrol_oneroster\OneRosterHelper;
+global $CFG;
```
Comment (Peterburnett) [inline tests/processcsv_filecheck_test.php:19]: this isnt required, as its called FROM the global namespace

File: tests/processcsv_filecheck_test.php [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+            OneRosterHelper::delete_directory($this->testDir);
+        }
+    }
+
+    /**
+     * Test the check_manifest_and_files method.
+     */
+    public function testCheckManifestAndFiles_allFilesPresent()
+    {
+        $missing_files = OneRosterHelper::check_manifest_and_files($this->manifestPath, $this->testDir);
```
Comment (Peterburnett) [inline tests/processcsv_filecheck_test.php:96]: These are great, but its also worth testing each function individually

**Comment by Ruben-Cooper** [comment]
Changes completed and committed

**Pull request closed without merging**